### DISCUSSION TITLE
Android versioning in flutter using codemagic

### DISCUSSION THREAD

**Discussion by Navaneet00** [codemagic-ci-cd/codemagic-docs#2359] [Q&A] [answered]
How to update the version code and version name of the app? 
When codemagic updates version code and version name using the latest version from google play store or app store and incrementing it... I want to assign this version in the app directly.
  I want to show version name in my app in the settings page. Also can u please suggest other ways to set the latest version of app using codemagic?

**Comment by dtrdic** [comment, marked as answer]
hello @Navaneet00,
There are several approaches you can use for build versioning on Codemagic. One of the easiest ways to increment the application version on every build is by using the built-in environment variables that Codemagic exports during the build, such as the `PROJECT_BUILD_NUMBER` variable. This variable represents the current build number of your project and is automatically incremented with each new build of your app.

If you are building a Flutter app, your build command in your yaml file would look like this;
```
- name: Build AAB with Flutter
  script: |
     flutter build appbundle --release \
     --build-name=1.0.$PROJECT_BUILD_NUMBER \
     --build-number=$PROJECT_BUILD_NUMBER
```

If you want to use the latest version from the Google Play Store, your build command would look like this:
```
- name: Build AAB with Flutter
  script: |
    LATEST_GOOGLE_PLAY_BUILD_NUMBER=$(($(google-play get-latest-build-number --package-name "$PACKAGE_NAME" --tracks="$GOOGLE_PLAY_TRACK") + 1))
    if [ -z $LATEST_GOOGLE_PLAY_BUILD_NUMBER ]; then
      UPDATED_BUILD_NUMBER=$BUILD_NUMBER
    else
      UPDATED_BUILD_NUMBER=$(($LATEST_GOOGLE_PLAY_BUILD_NUMBER + 1))
    fi
    flutter build appbundle --release \
      --build-name=1.0.$UPDATED_BUILD_NUMBER \
      --build-number=$UPDATED_BUILD_NUMBER
```

You will also need to make changes in the build.gradle file so that the code dynamically adjusts the `versionCode` and `versionName` based on the latest build number from Google Play or the environment variable BUILD_NUMBER
```
android {
    def latestGooglePlayBuildNumber = Integer.valueOf(System.env.LATEST_GOOGLE_PLAY_BUILD_NUMBER ?: System.env.BUILD_NUMBER ?: 0)

    defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId "io.codemagic.test”
        // You can update the following values to match your application needs.
        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-gradle-build-configuration.
        minSdkVersion flutter.minSdkVersion
        targetSdkVersion flutter.targetSdkVersion
        versionCode latestGooglePlayBuildNumber + 1
        versionName "1.0." + String.valueOf(latestGooglePlayBuildNumber + 1)
    }
}
```
But please check [this documentation](https://docs.codemagic.io/knowledge-codemagic/build-versioning/) for more information regarding automatic versioning.
You can also check our [example templates](https://github.com/codemagic-ci-cd/codemagic-sample-projects/blob/main/flutter/flutter-android-and-ios-yaml-demo-project/codemagic.yaml#L29) that you can also reuse.

If you encounter any road blocks, let me know, share the build ID, and I'll assist you through the process

**Comment by dtrdic** [comment]
I am closing the thread due to inactivity. Let me know if you have any questions by replying to this thread or opening a new one.
